File: test/javascript/ad.js
// ad.js -- JS port of test/python/ad.py (subset).
// Covers forward/reverse AD primitives: jacobian, gradient, hessian
// on SX/MX, evaluated via Function::call against numeric inputs.

const path = require("path");
const fs = require("fs");

const wasmDir = path.resolve(__dirname, "../../build-wasm/swig/wasm-js");
const modulePath = path.join(wasmDir, "casadi.js");
if (!fs.existsSync(modulePath)) {
  console.error(`SKIP: ${modulePath} not built.`);
  process.exit(2);
}

function assertEqual(actual, expected, msg) {
  const same = (actual === expected) ||
    ((typeof actual === 'bigint' || typeof actual === 'number') &&
     (typeof expected === 'bigint' || typeof expected === 'number') &&
     BigInt(actual) === BigInt(expected));
  if (!same) throw new Error(`${msg}: expected ${expected}, got ${actual}`);
}

function assertArrayAlmostEqual(actual, expected, places, msg) {
  if (actual.length !== expected.length) {
    throw new Error(`${msg}: length expected ${expected.length}, got ${actual.length}`);
  }
  const tol = Math.pow(10, -places);
  for (let i = 0; i < actual.length; ++i) {
    if (Math.abs(actual[i] - expected[i]) > tol) {
      throw new Error(`${msg}[${i}]: expected ${expected[i]}, got ${actual[i]}`);
    }
  }
}

// ----- tests -----

function test_sx_jacobian_scalar(M) {
  // f(x) = sin(x), df/dx = cos(x). At x=1: cos(1).
  const x = M.SX.sym("x", 1n, 1n);
  const f_sym = M.sin(x);
  const df = M.jacobian(f_sym, x);
  const f = M.Function("df", [x], [df]);
  const out = f(M.DM(1));
  assertArrayAlmostEqual(out.nonzeros(), [Math.cos(1)], 8, "d(sin)/dx at 1");
}

function test_mx_jacobian_scalar(M) {
  const x = M.MX.sym("x", 1n, 1n);
  const f_sym = M.times(x, x);    // f(x) = x^2
  const df = M.jacobian(f_sym, x);
  const f = M.Function("df", [x], [df]);
  // df/dx = 2x.  At x=3, expect 6.
  const out = f(M.DM(3));
  assertArrayAlmostEqual(out.nonzeros(), [6], 8, "d(x^2)/dx at 3");
}

function test_sx_gradient_vector(M) {
  // f(x, y) = x^2 + 3*x*y + y^2.  grad = [2x + 3y, 3x + 2y].
  const x = M.SX.sym("x");
  const y = M.SX.sym("y");
  const f_sym = M.plus(
      M.plus(M.times(x, x), M.times(M.times(M.SX(3), x), y)),
      M.times(y, y));
  const gx = M.gradient(f_sym, x);
  const gy = M.gradient(f_sym, y);
  const f = M.Function("g", [x, y], [gx, gy]);
  // At (x=2, y=5): gx = 2*2 + 3*5 = 19; gy = 3*2 + 2*5 = 16.
  const outs = f(M.DM(2), M.DM(5));
  assertArrayAlmostEqual(outs[0].nonzeros(), [19], 8, "gx at (2,5)");
  assertArrayAlmostEqual(outs[1].nonzeros(), [16], 8, "gy at (2,5)");
}

function test_mx_jacobian_vector(M) {
  // f(x) = [x0*x1, x0+x1].  J = [[x1, x0], [1, 1]].
  // Build inputs as a 2-vector MX, output as a 2-vector MX.
  const x = M.MX.sym("x", 2n, 1n);
  const x0 = x.get(false, M.Slice(0n, false));
  const x1 = x.get(false, M.Slice(1n, false));
  const f_sym = M.vcat([M.times(x0, x1), M.plus(x0, x1)]);
  const J = M.jacobian(f_sym, x);
  assertEqual(J.size1(), 2n, "J size1");
  assertEqual(J.size2(), 2n, "J size2");

  const f = M.Function("J", [x], [J]);
  // At x = [3, 5]: J = [[5, 3], [1, 1]].
  // Column-major: [5, 1, 3, 1].
  const xv = M.DM([3, 5]);
  const out = f(xv);
  assertArrayAlmostEqual(out.nonzeros(), [5, 1, 3, 1], 8, "J at (3,5)");
}

function test_sx_jacobian_of_gradient(M) {
  // 2nd-derivative via jacobian-of-gradient (avoids hessian's
  // [H, g] dual output, which currently has marshalling issues).
  // f(x) = x^4.  d2f/dx2 = 12*x^2.  At x=2: 48.
  const x = M.SX.sym("x");
  const f_sym = M.power(x, M.SX(4));
  const g = M.gradient(f_sym, x);
  const H = M.jacobian(g, x);
  const f = M.Function("H", [x], [H]);
  const out = f(M.DM(2));
  assertArrayAlmostEqual(out.nonzeros(), [48], 8, "d2(x^4)/dx2 at 2");
}

// ----- SX-vs-MX equivalence: the basic checkfunction_light pattern.
// Builds the same expression in both SX and MX, then verifies they
// produce identical numerical output on the same DM input.  This is
// the bulk of the ad.py "test_*eval_*" pattern adapted to wasm-js.

const h = require("./_helpers");

function _build_pair(M, builder, sym_in, n) {
  // Build the same Function in SX (with M.SX.sym) and MX (with M.MX.sym),
  // each of vector size `n`.  `builder(M, x)` returns the output expression.
  const xs = M.SX.sym("x", BigInt(n));
  const xm = M.MX.sym("x", BigInt(n));
  return [
    M.Function("fs", [xs], [builder(M, xs)]),
    M.Function("fm", [xm], [builder(M, xm)]),
  ];
}

function test_ad_equiv_x_squared(M) {
  // f(x) = x^2 elementwise, on a length-3 input.
  const [fs, fm] = _build_pair(M, (M, x) => M.times(x, x), null, 3);
  h.checkfunction_light(M, fs, fm, [M.DM([1.5, -2, 0.5])], 10);
}

function test_ad_equiv_trig_sum(M) {
  // f(x) = sin(x) + cos(x).
  const [fs, fm] = _build_pair(M, (M, x) => M.plus(M.sin(x), M.cos(x)), null, 4);
  h.checkfunction_light(M, fs, fm, [M.DM([0, 0.7, 1.4, 2.1])], 10);
}

function test_ad_equiv_dot_product(M) {
  // f(x) = dot(x, x).  SX vs MX should agree numerically.
  const builder = (M, x) => M.dot(x, x);
  const [fs, fm] = _build_pair(M, builder, null, 5);
  h.checkfunction_light(M, fs, fm, [M.DM([1, 2, 3, 4, 5])], 10);
}

function test_ad_equiv_jacobian(M) {
  // jacobian(sin(x), x) should agree as a Function between SX and MX.
  const xs = M.SX.sym("x", 3n);
  const fs = M.Function("Js", [xs], [M.jacobian(M.sin(xs), xs)]);
  const xm = M.MX.sym("x", 3n);
  const fm = M.Function("Jm", [xm], [M.jacobian(M.sin(xm), xm)]);
  h.checkfunction_light(M, fs, fm, [M.DM([0.1, 0.4, 0.9])], 10);
}

// ----- driver -----

const tests = [
  ["test_sx_jacobian_scalar",  test_sx_jacobian_scalar],
  ["test_mx_jacobian_scalar",  test_mx_jacobian_scalar],
  ["test_sx_gradient_vector",  test_sx_gradient_vector],
  ["test_mx_jacobian_vector",  test_mx_jacobian_vector],
  ["test_sx_jacobian_of_gradient", test_sx_jacobian_of_gradient],
  ["test_ad_equiv_x_squared",  test_ad_equiv_x_squared],
  ["test_ad_equiv_trig_sum",   test_ad_equiv_trig_sum],
  ["test_ad_equiv_dot_product", test_ad_equiv_dot_product],
  ["test_ad_equiv_jacobian",   test_ad_equiv_jacobian],
];

(async () => {
  const create = require(modulePath);
  const M = await create();
  let pass = 0, fail = 0;
  for (const [name, fn] of tests) {
    try { fn(M); console.log(`ok -- ${name}`); pass++; }
    catch (e) { console.log(`FAIL ${name}: ${e.message}`); fail++; }
  }
  console.log(`\n${pass}/${pass + fail} passed`);
  process.exit(fail === 0 ? 0 : 1);
})().catch((e) => { console.error("FATAL:", e); process.exit(1); });

File: test/julia/ad.jl
# ad.jl -- capability port of test/javascript/ad.js (+ deeper).
# jacobian / gradient / hessian / jtimes / forward / reverse, Jacobian
# sparsity, hessian-of-Lagrangian, all checked against analytic values.

const _AD_OD = Dict{String,Any}()

@testset "jacobian: scalar SX and MX" begin
    xs = sym(SX, "x")
    fs = C.Function("fs", [xs], [jacobian(sin(xs), xs)])
    @test abs(Float64(fs(DM(1.0))) - cos(1.0)) < 1e-10
    xm = sym(MX, "x")
    fm = C.Function("fm", [xm], [jacobian(C.times(xm, xm), xm)])
    @test abs(Float64(fm(DM(3.0))) - 6.0) < 1e-10   # d(x^2)/dx = 2x
end

@testset "jacobian: vector, sparsity + values" begin
    x = sym(SX, "x", 2)
    e = vertcat(C.times(x[1], x[2]), x[1] + x[2])     # f = [x0*x1, x0+x1]
    J = jacobian(e, x)
    @test Base.size(J) == (2, 2)
    # J = [[x1, x0], [1, 1]]; at x=[3,5] column-major nonzeros [5,1,3,1]
    f = C.Function("J", [x], [J])
    @test Vector{Float64}(f(DM([3.0, 5.0]))) == [5.0, 1.0, 3.0, 1.0]
    # Diagonal Jacobian has a sparse pattern (3 nnz on a 3x3).
    y = sym(SX, "y", 3)
    Jd = jacobian(C.times(y, y), y)
    @test C.nnz(C.sparsity(Jd)) == 3
end

@testset "gradient: analytic check" begin
    x = sym(SX, "x"); y = sym(SX, "y")
    f = C.times(x, x) + C.times(C.times(SX(3.0), x), y) + C.times(y, y)
    z = vertcat(x, y)
    g = gradient(f, z)                                # [2x+3y, 3x+2y]
    @test Base.size(g) == (2, 1)
    gf = C.Function("g", [z], [g])
    @test Vector{Float64}(gf(DM([2.0, 5.0]))) == [2 * 2 + 3 * 5, 3 * 2 + 2 * 5]
end

@testset "hessian: 2nd derivatives" begin
    x = sym(SX, "x", 2)
    f = C.times(x[1], x[1]) + C.times(C.times(x[1], x[2]), x[2])  # x0^2 + x0*x1^2
    H = hessian(f, x)
    @test Base.size(H) == (2, 2)
    hf = C.Function("H", [x], [H])
    # H = [[2, 2*x1], [2*x1, 2*x0]]; at x=[3,5] -> [[2,10],[10,6]]
    @test Matrix{Float64}(hf(DM([3.0, 5.0]))) == [2.0 10.0; 10.0 6.0]
    # x^4 -> d2/dx2 = 12 x^2 ; at 2 -> 48
    s = sym(SX, "s")
    h2 = C.Function("h2", [s], [hessian(s^4, s)])
    @test abs(Float64(h2(DM(2.0))) - 48.0) < 1e-9
end

@testset "jtimes: forward jacobian-vector product" begin
    y = sym(SX, "y", 2)
    e = vertcat(C.times(y[1], y[2]), y[1] + y[2])     # J = [[y1,y0],[1,1]]
    v = sym(SX, "v", 2)
    jt = C.jtimes(e, y, v, false, _AD_OD)             # J * v
    f = C.Function("jt", [y, v], [jt])
    # at y=[3,5], v=[1,0] -> [y1*1, 1*1] = [5, 1]
    @test Vector{Float64}(f(DM([3.0, 5.0]), DM([1.0, 0.0]))) == [5.0, 1.0]
    # at v=[0,1] -> [y0, 1] = [3, 1]
    @test Vector{Float64}(f(DM([3.0, 5.0]), DM([0.0, 1.0]))) == [3.0, 1.0]
end

@testset "forward / reverse directional derivatives" begin
    x = sym(SX, "x")
    f = C.Function("f", [x], [C.times(x, x)])     # f(x) = x^2
    # forward: inputs (x, nominal_out, seed) -> 2*x*seed
    fwd = C.forward(f, 1)
    @test C.n_in(fwd) == 3
    @test abs(Float64(fwd(DM(3.0), DM(9.0), DM(1.0))) - 6.0) < 1e-10
    @test abs(Float64(fwd(DM(3.0), DM(9.0), DM(2.0))) - 12.0) < 1e-10
    # reverse: same chain rule, adjoint seed -> 2*x*seed
    rev = C.reverse(f, 1)
    @test C.n_in(rev) == 3
    @test abs(Float64(rev(DM(3.0), DM(9.0), DM(1.0))) - 6.0) < 1e-10
end

@testset "hessian of the Lagrangian via factory" begin
    x = sym(SX, "x", 2); p = sym(SX, "p")
    nlp = Dict("x" => x, "p" => p, "f" => x[1]^2 + C.times(p, x[2]^2), "g" => x[1] + x[2])
    f = C.Function("f", nlp, ["x", "p"], ["f", "g"], _AD_OD)
    H = C.factory(f, "H", ["x", "p"], ["hess:f:x:x"],
        Dict{String,Vector{String}}(), _AD_OD)
    @test C.n_out(H) == 1
    # hess(x0^2 + p*x1^2) wrt x = diag(2, 2p); at p=3 -> diag(2,6)
    @test Matrix{Float64}(H(DM([1.0, 1.0]), DM(3.0))) == [2.0 0.0; 0.0 6.0]
end

@testset "SX vs MX equivalence on shared expressions" begin
    build(x) = C.sin(x) + C.cos(x)
    xs = sym(SX, "x", 4); xm = sym(MX, "x", 4)
    fs = C.Function("fs", [xs], [build(xs)])
    fm = C.Function("fm", [xm], [build(xm)])
    inp = DM([0.0, 0.7, 1.4, 2.1])
    @test maximum(abs.(Vector{Float64}(fs(inp)) .- Vector{Float64}(fm(inp)))) < 1e-12
    # jacobians agree too
    Js = C.Function("Js", [xs], [jacobian(build(xs), xs)])
    Jm = C.Function("Jm", [xm], [jacobian(build(xm), xm)])
    @test maximum(abs.(Vector{Float64}(Js(inp)) .- Vector{Float64}(Jm(inp)))) < 1e-12
end
